src/s3/document/s3DocumentProvider.ts

```diff
@@ -0,0 +1,27 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { ext } from '../../shared/extensionGlobals'
+import { readFileAsString } from '../../shared/filesystemUtilities'
+import { showOutputMessage } from '../../shared/utilities/messages'
+
+export class S3DocumentProvider implements vscode.TextDocumentContentProvider {
+    public constructor() {}
+
+    onDidChange?: vscode.Event<vscode.Uri> | undefined
+
+    public async provideTextDocumentContent(uri: vscode.Uri): Promise<string> {
+        let data: any
+        try {
+            //data = await fs.readFileSync(uri.fsPath)
+            data = await readFileAsString(uri.fsPath)
+        } catch (e) {
+            showOutputMessage(`${e}`, ext.outputChannel)
+        }
+
+        return data ?? ''
```

Comment:
Do you mean when data is `undefined` ? What can this be the case? IMO it it only when `uri.fsPath` was not found, and because of the fileViewerManager it can be assumed that it will be there. However, I see your point in the edge case that this doesn't read anything, what do you suggest?

Reply:
An explicit error would be better than silently opening an empty text editor.

Reply:
I am not sure how the case when it returns empty is undefined, since vscode is the one that calls and uses this method

Reply:
Well, this is returning an explicit empty string so I'd expect that would open an empty virtual document but you can test that locally. We may feel that getting in to this case is unlikely (since the `fileViewerManager` is assumed to assure the file exists) but it's always good to write defensive code and avoid silently failing.

Reply:
Addressed: [https://github.com/aws/aws-toolkit-vscode/pull/1878/files#diff-282b2f17246da4dcf9290625ed89c103b751fbd833b018a9eb2c213bb85800bdR24-R26](PR1878)